### PR TITLE
Add "Colors" extension

### PR DESCRIPTION
**Pull request by linuz90**
## Description

This extension uses the `tinycolor` library to convert your provided color into various formats. It's very handy for designers and developers that need to do quick conversions.

For example, I often prefer HSL to RGB or HEX since it lets me change saturation and luminance more easily.

You can press enter on a converted color to copy it, or Command + Enter to insert it inline.

## Type of change

- New extension

## Screencast

https://user-images.githubusercontent.com/9629188/151674382-e431c38a-204b-4d36-b2ab-045cf14b6979.mp4

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

**Comment by linuz90** [comment]
Forgot to showcase that invalid colors will throw an error:

CleanShot 2022-01-29 at 19 17 45@2x

File: extensions/colors/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  CopyToClipboardAction,
+  Form,
+  Icon,
+  List,
+  PasteAction,
+  SubmitFormAction,
+  Toast,
+  ToastStyle,
+  useNavigation,
+} from "@raycast/api";
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:13]:
```suggestion
import {
  ActionPanel,
  ActionPanelItem,
  CopyToClipboardAction,
  Form,
  Icon,
  List,
  PasteAction,
  showToast,
  SubmitFormAction,
  ToastStyle,
  useNavigation,
} from "@raycast/api";
```

File: extensions/colors/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+export default function Command() {
+  const navigation = useNavigation();
+
+  function handleSubmit(form: { color: string }) {
+    const color = form.color;
+
+    const toast = new Toast({ title: "", style: ToastStyle.Failure });
+    toast.hide();
+
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:24]:
```suggestion
```

File: extensions/colors/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    if (!color) {
+      toast.title = "Please enter a color";
+      toast.show();
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:27]:
```suggestion
      showToast(ToastStyle.Failure, "Please enter a color");
```

File: extensions/colors/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      return;
+    }
+    const parsedColor = tinycolor(color);
+    if (!parsedColor.isValid()) {
+      toast.title = "This is not a valid color";
+      toast.show();
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:33]:
```suggestion
      showToast(ToastStyle.Failure, "This is not a valid color");
```

File: extensions/colors/src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      return;
+    }
+
+    navigation.push(<ColorView parsedColor={parsedColor} />);
+  }
+
+  return (
+    <Form
+      navigationTitle="Insert color to convert"
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:42]:
```suggestion
      navigationTitle="Insert Color to Convert"
```

File: extensions/colors/src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  );
+}
+
+const ColorView = ({ parsedColor: parsed }: { parsedColor: tinycolor.Instance }) => {
+  const navigation = useNavigation();
+
+  const c = parsed.toHexString();
+
+  return (
+    <List navigationTitle="Converted colors">
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:60]:
```suggestion
    <List>
```
Redundant

File: extensions/colors/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      <List.Section>
+        <List.Item
+          title="Convert new color"
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:63]:
```suggestion
          title="Convert New Color"
```

File: extensions/colors/src/index.tsx [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+          icon={Icon.ArrowClockwise}
+          actions={
+            <ActionPanel>
+              <ActionPanelItem title="New color" onAction={() => navigation.pop()} />
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:67]:
```suggestion
              <ActionPanelItem title="New Color" onAction={() => navigation.pop()} />
```

File: extensions/colors/src/index.tsx [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            </ActionPanel>
+          }
+        />
+      </List.Section>
+      <List.Section title="Converted colors">
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:72]:
```suggestion
      <List.Section title="Converted Colors">
```

File: extensions/colors/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        <ColorItem description="HEX" color={c} value={parsed.toHexString()} />
+        {parsed.getAlpha() < 1 && <ColorItem description="HEX8" color={c} value={parsed.toHex8String()} />}
+        <ColorItem description="RGB" color={c} value={parsed.toRgbString()} />
+        <ColorItem description="HSL" color={c} value={parsed.toHslString()} />
+        <ColorItem description="HSV" color={c} value={parsed.toHsvString()} />
+        <ColorItem
+          description="Perceived brightness"
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:79]:
```suggestion
          description="Perceived Brightness"
```

File: extensions/colors/src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+      accessoryTitle={description}
+      actions={<ColorActions content={value} />}
+    />
+  );
+};
+
+const ColorActions = ({ content }: { content: string }) => {
+  return (
+    <ActionPanel>
+      <CopyToClipboardAction title="Copy to clipboard" content={content} />
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:106]:
```suggestion
      <CopyToClipboardAction title="Copy to Clipboard" content={content} />
```

File: extensions/colors/src/index.tsx [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+      <PasteAction title="Insert" content={content} shortcut={{ modifiers: ["cmd"], key: "enter" }} />
```
Comment (pernielsentikaer) [inline extensions/colors/src/index.tsx:107]:
```suggestion
      <PasteAction title="Paste" content={content} shortcut={{ modifiers: ["cmd"], key: "enter" }} />
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @linuz90 👋 

Thanks for your contribution 💪 

I left some suggestions for you to work with, also, it could be cool with:
- A way to select default action, if copy or paste should be default ([Like done in this extension](https://github.com/raycast/extensions/pull/593/files))

- A new command that takes selection as color value (idea)

- A new command that takes clipboard as color value (idea)

Looking forward to testing this out again 💪 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by fe9lix** [comment]
Hey @linuz90, what's the status of this PR? Are you planning on addressing the feedback from @pernielsentikaer?

**Comment by linuz90** [comment]
> Hey @linuz90, what's the status of this PR? Are you planning on addressing the feedback from @pernielsentikaer?

Hey, I'd love to but I'm just too busy working on Typefully now, so I'll likely won't have time for this for a few weeks.

**Comment by fe9lix** [comment]
Are you fine with closing this for now and then re-opening a new PR with the feedback addressed, once you've found time?

**Comment by linuz90** [comment]
Sure, no problem!

**Comment by pernielsentikaer** [comment]
Closing as agreed

**Pull request closed without merging**